### ISSUE TITLE
Support Medusa Sampling

### ISSUE THREAD

**Issue by forpanyang** [NVIDIA/TensorRT-LLM#25]
Mirror issue: https://github.com/vllm-project/vllm/issues/1023

**Comment by jdemouth** [comment]
We are considering different techniques of "speculative sampling/decoding" for future releases. We are currently defining what will be part of the next release (expected to be released in roughly a month) and we will let you know in a couple of weeks if that feature will be part of it.

**Comment by LiuXiaoxuanPKU** [comment]
Thanks for the great project!
I feel the [standard speculative decoding](https://proceedings.mlr.press/v202/leviathan23a.html#:~:text=Inference%20from%20large%20autoregressive%20models,computing%20several%20tokens%20in%20parallel.)  is a must-have feature. It would be much easier if tensorRT could provide hf-like forward API, where developers can access logits and past kvs. I notice the function [here](https://github.com/NVIDIA/TensorRT-LLM/blob/1c4a0ee24cb0480c20a721c4bc6a77fc51644a22/tensorrt_llm/runtime/generation.py#L1381), but not sure if that's the correct start point. Any comments are highly appreciated and I am happy to contribute.

**Comment by jdemouth-nvidia** [comment]
Hi @LiuXiaoxuanPKU , thanks for your message. We are, indeed, working on speculative decoding, including the standard version. We hope to have something working in a few weeks. 

For the logits, you can get them from the forward pass, the `logits` tensor is exposed and populated by the `forward` pass through the TensorRT engine. You can also get the context logits (i.e. not just the last token in the context phase). For the past K/V values, we have a K/V cache that gets populated as we run. If you tell us what you would need precisely, we can think of a way to expose an API to get access to the data.

Thanks,
Julien

**Comment by LiuXiaoxuanPKU** [comment]
Thanks for the reply, to implement standard speculative decoding, I only need:
(1) logits: seems no problem here.
(2) truncate kv cache: for example, given sequence of tokens, including prompt and generated tokens [1, 2, 3, 4, 5], it turns out 4 and 5 are incorrect. Therefore, the kv cache of 4, 5 should be invalidated.
(3) forward function given kv cache: Still the example above,  the input of the forward function will be the kv cache of [1, 2, 3] and a new token provided by the target model [7] (i.e., we need to restart from [1,2,3,7] and continue the generation.) 

Also, on a high level, do you suggest implementing it in the python runtime or cpp runtime?

**Comment by LiuXiaoxuanPKU** [comment]
Hi, I am playing with the python runtime and trying to get the context_logits, but I feel it cannot fulfill the requirements.

Concretely, I can get the logits of generated tokens
```
prompt: [1, 2, 3, 4]
generated tokens: [5, 6, 7] # I can get the logits for token 5, 6, 7.
```

In speculative decoding, I need the logits of the 'prompt', for example
```
prompt: [1, 2, 3, 4] # I need the logits for token 3, 4 if they are tokens proposed by the draft model
generated tokens: [5]
```
More precisely, I need the logits of token 3 and 4 when generating token 5, I should get those information in a single forward pass. Please let me know if there are ways to achieve this in the python runtime, thanks a lot!

**Comment by jdemouth-nvidia** [comment]
Hi @LiuXiaoxuanPKU ,

Thanks a lot for looking into this.

To answer the question regarding Python or C++ runtimes, we are currently focusing on the C++ runtime in the TensorRT-LLM team. We want to offer great performance and a lot of features in C++ first. Then, we offer a proper binding for Python. We already have partners/customers/friends working on a binding of the GptManager: https://github.com/NVIDIA/TensorRT-LLM/pull/76.

Regarding the context logits, did you take a look at `gather_all_token_logits` as in that [code](https://github.com/NVIDIA/TensorRT-LLM/blob/1c4a0ee24cb0480c20a721c4bc6a77fc51644a22/tensorrt_llm/runtime/generation.py#L1458). If you enable it, it'll return all the logits and not just the logits of the last token. Isn't it what you need?

Thanks,
Julien

**Comment by LiuXiaoxuanPKU** [comment]
Thanks for the information. Yeah works for me, thanks.

**Comment by Dev-hestabit** [comment]
@LiuXiaoxuanPKU are you able to implement speculative decoding in tensorRT LLM

**Comment by zhyncs** [comment]
> We are considering different techniques of "speculative sampling/decoding" for future releases. We are currently defining what will be part of the next release (expected to be released in roughly a month) and we will let you know in a couple of weeks if that feature will be part of it.

Hi @jdemouth @Shixiaowei02 @juney-nvidia Do you have any plans to support [SafeAILab/EAGLE](https://github.com/SafeAILab/EAGLE)?

**Comment by aspctu** [comment]
@ncomly-nvidia Bump on the above, any plans to support EAGLE?
